drivers/SmartThings/zigbee-window-treatment/src/aqara/init.lua

```diff
@@ -0,0 +1,275 @@
+  local state = value.value
+  device:set_field(SHADE_STATE, state)
+
+  if state == SHADE_STATE_STOP then
+    local shadeLevel = device:get_field(SHADE_LEVEL) or 0
+    emit_level_event(device, shadeLevel)
+
+    local flag = device:get_field(INIT_STATE) or INIT_STATE_DONE
+    if flag == INIT_STATE_CLOSE then
+      device:set_field(INIT_STATE, INIT_STATE_REVERSE)
+      send_open_cmd(device, "main")
+    elseif flag == INIT_STATE_OPEN then
+      device:set_field(INIT_STATE, INIT_STATE_REVERSE)
+      send_close_cmd(device, "main")
+    elseif flag == INIT_STATE_REVERSE then
+      device:set_field(INIT_STATE, INIT_STATE_DONE)
+      read_pref_attribute(device)
+    end
+  elseif state == SHADE_STATE_OPEN then
+    device:emit_event(capabilities.windowShade.windowShade.opening())
+  elseif state == SHADE_STATE_CLOSE then
+    device:emit_event(capabilities.windowShade.windowShade.closing())
+  end
+end
+
+local function current_position_attr_handler(driver, device, value, zb_rx)
+  local level = value.value
+  if level == 5.8774717541114e-39 then
+    -- float 5.8774717541114e-39 is 0
+    level = 0
+  end
+  if level > 100 then
+    level = 100
+  end
+  level = utils.round(level)
+
+  device:set_field(SHADE_LEVEL, level)
+  device:emit_event(capabilities.windowShadeLevel.shadeLevel(level))
+
+  local shadeState = device:get_field(SHADE_STATE) or SHADE_STATE_STOP
+  if shadeState == SHADE_STATE_STOP then
+    emit_level_event(device, level)
+  end
+end
+
+local function pref_attr_handler(driver, device, value, zb_rx)
+  local initialized = string.byte(value.value, 3) & 0xFF
+  local flag = device:get_field(INIT_STATE) or INIT_STATE_DONE
+  if flag == INIT_STATE_DONE then
+    device:emit_event(initialized == 1 and initializedstate.initializedState.initialized() or
+      initializedstate.initializedState.notInitialized())
+  end
+end
+
+local aqara_window_treatment_handler = {
+  NAME = "Aqara Window Treatment Handler",
+  lifecycle_handlers = {
+    added = device_added,
+    doConfigure = do_configure,
+    infoChanged = device_info_changed
+  },
+  capability_handlers = {
+    [capabilities.windowShadeLevel.ID] = {
+      [capabilities.windowShadeLevel.commands.setShadeLevel.NAME] = window_shade_level_cmd
+    },
+    [capabilities.windowShade.ID] = {
+      [capabilities.windowShade.commands.open.NAME] = window_shade_open_cmd,
+      [capabilities.windowShade.commands.close.NAME] = window_shade_close_cmd,
+      [capabilities.windowShade.commands.pause.NAME] = window_shade_pause_cmd
+    },
+    [capabilities.refresh.ID] = {
+      [capabilities.refresh.commands.refresh.NAME] = do_refresh,
+    },
+    [initializedStateId] = {
+      [setInitializedStateCommandName] = setInitializedState_handler
+    }
+  },
+  zigbee_handlers = {
+    attr = {
+      [WindowCovering.ID] = {
+        [WindowCovering.attributes.CurrentPositionLiftPercentage.ID] = current_position_attr_handler

```

Comment:
Ah, this is what was confusing me is that you have this handler registered for both this attribute and the analog output cluster below.  So is it the case that the device doesn't report this value?  If it reports both, it seems like we should just use this value as it is closer to the value we need.

Reply:
The AnalogOutput.attributes.PresentValue is closer to the value. I updated It.